compress.cc

```diff
@@ -110,7 +114,26 @@ compressor::ptr_type compressor::create(const sstring& name, const opt_getter& o
         }
     }
 
-    return compressor_registry::create(qn, opts);
+    static thread_local std::unordered_map<sstring, ptr_type> named_compressors;
+
+    ptr_type res = nullptr;
+
+    auto i = named_compressors.find(name);
+    if (i != named_compressors.end()) {
+        res = i->second;
+        auto rep = res->replace(opts);
+        if (res) {
+            res = std::move(rep);
+        }
+    }
+
+    if (!res) {
+        res = compressor_registry::create(qn, opts);
+        assert(res);
```

Comment:
Please use `on_internal_error` instead

Reply:
See https://github.com/scylladb/scylladb/issues/7871#issuecomment-1427680016

Reply:
I'm not sure if you disagree or not:

> Actual, runtime, checks for broken/uncontinuable conditions should as you say be various forms of ensure/internal_error

This looks like a runtime check for a broken/uncontinuable condition.
And AFAIU the point of issue #7871 is to use `on_internal_error` everywhere where you'd use `assert`.

Reply:
IF so, then it is again breaking the usage of assert. `compressor_registry::create` is not supposed to be able to return nullptr. It should only return non-null or throw. A runtime check `if (...) on_internal_error();` suggests it _can_ and we are doing unconditional checks. An assert (when used correctly) is a check only enabled in debug/assert builds, that add the extra overhead of sometimes elaborate checks to verify assumptions more thoroughly. Yet, once internal testing is done, and we assume that broken API:s etc are fixed, they can be compiled away. This is how asserts should be used. The way scylla uses them (always on, glorified "if() throw") is imho not good. Which is why I support https://github.com/scylladb/scylladb/issues/7871 in converting things that _are_ always check into what you say, but, as my comment tried to entail, also move actual asserts to _actual_ asserts (as god intended them)

Reply:
Yes, it is functionally an unconditional assert. But is looks like a runtime check. And it does not disappear when the enable assert preprocessor macro is 0. And it lacks the very distinct uniqueness of `assert`, which tells the reader quickly "here be a sanity check that should not be needed".

Reply:
@avikivity Why did Scylla/Seastar make the decision to have no development-only asserts? This only leads us to limit the number of invariant checks we do (due to performance reasons), which (probably) leads to longer bug investigations. Why not have two types of assert — one for expensive development-only checks and one for release checks?